felix/iptables/match_builder.go

```diff
@@ -343,3 +347,29 @@ func PortRangessToMultiport(ports []*proto.PortRange) string {
 	portsString := strings.Join(portFragments, ",")
 	return portsString
 }
+
+// vniMatchBPF returns a cBPF socket filter program which matches on the VXLAN
+// Network ID of encapsulated packets. The program assumes that it will only be
+// applied to VXLAN UDP datagrams. It is compatible with both IPv4 and IPv6.
+func matchVXLANVNI(vni uint32) []bpf.RawInstruction {
+	asm, err := bpf.Assemble([]bpf.Instruction{
+		// Load offset of UDP payload into X.
+		bpf.LoadExtension{Num: bpf.ExtPayloadOffset}, // ld poff
+		bpf.TAX{}, // tax
+
+		bpf.LoadIndirect{Off: 4, Size: 4},                      // ld [x + 4] ; Load VXLAN ID into top 24 bits of A
+		bpf.ALUOpConstant{Op: bpf.ALUOpShiftRight, Val: 8},     // rsh #8     ; A >>= 8
+		bpf.JumpIf{Cond: bpf.JumpEqual, Val: vni, SkipTrue: 1}, // jeq $vni, match
+		bpf.RetConstant{Val: 0},                                // ret #0
+		bpf.RetConstant{Val: ^uint32(0)},                       // match: ret #-1
+	})
+	// bpf.Assemble() only errors if an instruction is invalid. As the only variable
+	// part of the program is an instruction value for which the entire range is
+	// valid, whether the program can be successfully assembled is independent of
+	// the input. Given that the only recourse is to fix this function and
+	// recompile, there's little value in bubbling the error up to the caller.
+	if err != nil {
+		panic(err)
```

Comment:
```suggestion
		log.WithError(err).Panic("Failed to assemble static BPF VNI match program.")
```
Agree a panic is reasonable here.  Please panic via the logging library; it flushes the log on the way down.